docs/design/daemon.md

```diff
@@ -0,0 +1,128 @@
+
+## Design
+
+#### Client
+- Add support for daemon controller commands to kubectl and the client. Client code was added to client/unversioned. The main files in Kubectl that were modified are kubectl/describe.go and kubectl/stop.go, since for other calls like Get, Create, and Update, the client simply forwards the request to the backend via the REST API.
+
+#### Apiserver
+- Accept, parse, validate client commands
+- REST API calls are handled in registry/daemon
+  - In particular, the api server will add the object to etcd
+  - DaemonManager listens for updates to etcd (using Framework.informer)
+- API objects for Daemon Controller were created in expapi/v1/types.go and expapi/v1/register.go
+- Validation code is in expapi/validation
+
+#### Daemon Manager
+- Creates new daemon controllers when requested. Launches the corresponding daemon pod on all nodes with labels matching the new daemon controller’s selector.
+- Listens for addition of new nodes to the cluster, by setting up a framework.NewInformer that watches for the creation of Node API objects. When a new node is added, the daemon manager will loop through each daemon controller. If the label of the node matches the selector of the daemon controller, then the daemon manager will create the corresponding daemon pod in the new node.
+- The daemon manager creates a pod on a node by sending a command to the API server, requesting for a pod to be bound to the node (the node will be specified via its hostname)
+
+#### Kubelet
+- Does not need to be modified, but health checking will occur for the daemon pods and revive the pods if they are killed (we set the pod restartPolicy to Always). We reject Daemon Controller objects with pod templates that don’t have restartPolicy set to Always.
+
+## Testing
+
+Unit Tests:
+Each component was unit tested, fakes were implemented when necessary. For example, when testing the client, a fake API server was used.
+
+End to End Tests:
+One end-to-end test was implemented. The end-to-end test verified that the daemon manager runs the daemon on every node, that when a daemon pod is stopped it restarts, that the daemon controller can be reaped (stopped), and that the daemon adds/removes daemon pods appropriately from nodes when their labels change.
```

Comment:
No, but it does flip the label on a node from not-matching to matching and verifies that a daemon is launched. I'll file an issue for the test you described (#14314).